// include/v2/connect.js
/*
 * small module that connects to a MySQL db
 * credentials should be stored only in this place
 *
 * version 2 - using events only
 * */
const credentials = {
    host: "localhost",
    user: "notes_user",
    password: "notes_password",
    db: "notes"
};

const mysql = require("mysql");

function connect_to_db(ee, event, args){
    /*
     * connects to mysql db using global var "credentials"
     * "credentials" and this function should be kept in same file
     *
     * ee: event emitter that listens for event
     * callback params: (ret, args)
     *
     * args: args to be passed to event callback
     *   too generic, I don't need this with v2
     *
     * event: event to be emitted on ee when done
     *
     * ret {
     *   "success": boolean,
     *   "messages": [string, ...], (array of strings)
     *   "connection": mysql.Connection object (optional, in case of success)
     * }
     */
    var connection = mysql.createConnection({
	host: credentials.host,
	user: credentials.user,
	password: credentials.password,
	database: credentials.db
    });
    connection.connect(function(err){
	if (err){
	    let ret = {
		success: false,
		messages: ["Cannot connect to db: " + err.message],
	    }
	    ee.emit(event, ret, args);
	}
	let ret = {
	    success: true,
	    messages: ["success"],
	    connection: connection,
	};
	ee.emit(event, ret, args);
    });
}

/* exports */
module.exports.connect_to_db = connect_to_db;
