// include/v1/connect.js
/*
 * small module that connects to a MySQL db
 * credentials should be stored only in this place
 *
 * version 1
 * this is probably not the best implementation (returning connection)
 * check version 2 for possibly a better version
 * */
const credentials = {
    host: "localhost",
    user: "notes_user",
    password: "notes_password",
    db: "notes"
};

const mysql = require("mysql");

function connect_to_db(ee, event, args){
    /*
     * connects to mysql db using global var "credentials"
     * "credentials" and this function should be kept in same file
     *
     * ee: event emitter that listens for event
     * callback params: (ret, args)
     *
     * args: args to be passed to event callback
     *
     * returns mysql.Connection object
     *
     * if anything goes bad it will emit an event on ee
     * passing ({success: false, messages:...}, args)
     */
    var connection = mysql.createConnection({
	host: credentials.host,
	user: credentials.user,
	password: credentials.password,
	database: credentials.db
    });
    connection.connect(function(err){
	if (err){
	    let ret = {
		success: false,
		messages: ["Cannot connect to db: " + err.message],
	    }
	    ee.emit(event, ret, args);
	}
    });
    return connection;
}

/* exports */
module.exports.connect_to_db = connect_to_db;
